src/coprocessor/gadgets.rs

```diff
@@ -104,50 +114,425 @@ pub fn construct_cons<F: LurkField, CS: ConstraintSystem<F>>(
 }
 
 /// Constructs a cons-list with the provided `elts`. The terminating value defaults
-/// to the `nil` allocated pointer when `last` is `None`
+/// to `nil` when `last` is `None`
 pub fn construct_list<F: LurkField, CS: ConstraintSystem<F>>(
     cs: &mut CS,
     g: &GlobalAllocator<F>,
     store: &Store<F>,
     elts: &[&AllocatedPtr<F>],
     last: Option<AllocatedPtr<F>>,
 ) -> Result<AllocatedPtr<F>, SynthesisError> {
-    let init = last.unwrap_or_else(|| {
-        g.get_allocated_ptr_from_ptr(&store.intern_nil(), store)
-            .expect("nil pointer not allocated")
-    });
+    let init = match last {
+        Some(last) => last,
+        None => g.get_allocated_ptr_from_ptr(&store.intern_nil(), store)?,
+    };
     elts.iter()
         .rev()
         .enumerate()
         .try_fold(init, |acc, (i, ptr)| {
-            construct_cons(cs.namespace(|| format!("Cons {i}")), g, store, ptr, &acc)
+            construct_cons(cs.namespace(|| format!("cons {i}")), g, store, ptr, &acc)
+        })
+}
+
+/// Retrieves the `Ptr` that corresponds to `a_ptr` by using the `Store` as the
+/// hint provider
+fn get_ptr<F: LurkField>(
+    a_ptr: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<Ptr<F>, SynthesisError> {
+    let z_ptr = ZPtr::from_parts(
+        Tag::from_field(
+            &a_ptr
+                .tag()
+                .get_value()
+                .ok_or_else(|| SynthesisError::AssignmentMissing)?,
+        )
+        .ok_or_else(|| SynthesisError::Unsatisfiable)?,
+        a_ptr
+            .hash()
+            .get_value()
+            .ok_or_else(|| SynthesisError::AssignmentMissing)?,
+    );
+    Ok(store.to_ptr(&z_ptr))
+}
+
+/// Deconstructs `tuple`, assumed to be a composition of two others. Panics if
+/// the store can't deconstruct the tuple pointer
+pub fn deconstruct_tuple2<F: LurkField, CS: ConstraintSystem<F>>(
```

Comment:
https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components

```
/// Deconstructs `tuple`, assumed to be a composition of two others. 
/// # Panics:
/// if the store can't deconstruct the tuple pointer.